### ISSUE TITLE
yarn does not prefer latest releases in ranges like npm does

### ISSUE THREAD

**Issue by alangpierce** [yarnpkg/yarn#2988]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**

When installing a package at a version range, it looks like yarn's algorithm is to take the highest version matching that range, while npm's algorithm is similar, except that if any of those versions is tagged `latest`, it is used rather than the highest version.

This broke the decaffeinate project when installed through yarn (https://github.com/decaffeinate/decaffeinate/issues/931). `decaffeinate` depends on `decaffeinate-parser`, and `decaffeinate-parser` version 16.0.9 has a dependency on `decaffeinate-coffeescript` with range `^1.10.0-patch21`. The `decaffeinate-coffeescript` project has both `1.10.0` and `1.10.0-patch21`, with `1.10.0-patch21` marked `latest`. yarn was installing `1.10.0`, while npm was installing `1.10.0-patch21`. I can just change `decaffeinate-parser` to use an exact version instead of a range, so this isn't blocking me, but certainly it's a subtle issue.

I suspect this is the cause of #1545 as well.

I published a test package in the npm registry to exercise this case: `alangpierce-latest-test`. It has versions 1.0.0, 1.1.0, and 1.2.0, and version 1.1.0 is tagged "latest".

Here's the behavior of npm vs yarn:
```
> npm install alangpierce-latest-test
[installs alangpierce-latest-test@1.1.0]
> yarn add alangpierce-latest-test
[installs alangpierce-latest-test@1.1.0]
> npm install alangpierce-latest-test@^1.0.0
[installs alangpierce-latest-test@1.1.0]
> yarn add alangpierce-latest-test@^1.0.0
[installs alangpierce-latest-test@1.2.0]
```

**What is the expected behavior?**

```
> yarn add alangpierce-latest-test@^1.0.0
[installs alangpierce-latest-test@1.1.0]
```

**Please mention your node.js, yarn and operating system version.**

yarn 0.21.3
npm 3.10.10
node v6.9.4
macOS 10.12.3 (16D32)

**Comment by Daniel15** [comment]
> decaffeinate depends on decaffeinate-parser, and decaffeinate-parser version 16.0.9 has a dependency on decaffeinate-coffeescript with range ^1.10.0-patch21. The decaffeinate-coffeescript project has both 1.10.0 and 1.10.0-patch21, with 1.10.0-patch21 marked latest. yarn was installing 1.10.0, while npm was installing 1.10.0-patch21. I

`1.10.0-patch21` is older than `1.10.0` according to semver, so IMO Yarn's behaviour is correct and npm's behavior is anomalous. Having said that, Yarn should probably replicate npm's issues too, for compatibility.

**Comment by alangpierce** [comment]
Yep, agreed, probably it was a mistake to try to use pre-release versions to indicate patch numbers on top of the CoffeeScript version. Or at least I shouldn't have published a "base" 1.10.0 version. The fact that npm was relying on "latest" was a total accident. Now I know. :-)

I updated the decaffeinate-parser dependencies, so I'm not running into this problem anymore.

As for whether yarn should go with its more straightforward algorithm or copy npm's behavior, I could hear both sides. There may also be other options, like a compatibility mode flag or a warning when yarn detects that it will behave differently from npm. In my case, installing decaffeinate through yarn resulted in a tool that usually seemed to work, but would sometimes run into bugs because a number of bug fixes weren't applied. That's a particularly scary case because there wasn't anything obviously wrong (which is probably why it went months without anyone discovering it), but it might have ended up transforming code incorrectly.
